src/main/flight/position.c

```diff
@@ -106,126 +106,108 @@ void calculateEstimatedAltitude()
+            } else {
+                if (haveBaroAlt && haveGpsAlt) { // armed without a 3D fix, but have Baro, and later got a 3D fix
+                    // this will only be relevant for users without GPS Rescue enabled, or GPS Rescue in ignore 3D fix mode
+                    gpsAltOffsetCm = baroAltCm;
+                    okToZeroGps = true;
+                }
+            }
+        } else {
+            gpsAltOffsetCm = gpsAltCm; // while disarmed, keep capturing current altitude, to zero the GPS altitude once armed
+            altitudeRelativeCm = 0.0f; // and keep relative altitude at zero
+            gpsAltitudeZeroed = false; // won't be set true until arming
+            okToZeroGps = (haveGpsAlt); // true if we have a 3D fix just before arming, otherwise false
         }
     } else {
-        gpsAltOffsetCm = gpsAltCm; // while disarmed, keep capturing current altitude, to zero any offset once armed
-        altitudeToFilterCm = 0.0f; // while disarmed, force altitudeTofilter to zero, but leave gpsAltCm alone
-        altitudeOffsetSetGPS = (haveGpsAlt && gpsConfig()->gpsRequiredSats);
-        // if no 3D fix and not enough sats, set the GPS offset flag to false, use whatever offset we have, and later use baro to correct once we get a fix
+        gpsTrust = 0.0f; // don't trust GPS for altitude if no sensor or 3D fix
     }
 #endif
 
-    // *** adjust gpsTrust, favouring Baro increasingly when there is a discrepancy of more than a meter
-    // favour GPS if Baro reads negative, this happens due to ground effects
-    float gpsTrustModifier = gpsTrust;
-    const float absDifferenceM = fabsf(altitudeToFilterCm - baroAltCm) * positionConfig()->altitude_prefer_baro / 10000.0f;
-    if (absDifferenceM > 1.0f) { // when there is a large difference, favour Baro
-        // warning: "ground effect" in last 10cm of descent causes abrupt increase in pressure
-        // this is interpreted as an abrupt fall in altitude and could cause motors to increase just as we hit ground, leading to bouncing
-        gpsTrustModifier /=  absDifferenceM;
-    }
-    // eg if discrepancy is 3m and GPS trust was 0.9, it would now be 0.3
-
-    if (haveGpsAlt && !(positionConfig()->altitude_source == BARO_ONLY)) {
-        if (haveBaroAlt) { // have a baro signal, and are in GPS_ONLY or DEFAULT modes, so mix GPS and Baro
-            altitudeToFilterCm = altitudeToFilterCm * gpsTrustModifier + baroAltCm * (1 - gpsTrustModifier);
-            // if we have no Baro signal, then altitudeToFilterCm and baroAltCm carry existing GPS altitude values through
+    if (gpsAltitudeZeroed && (positionConfig()->altitude_source == DEFAULT || positionConfig()->altitude_source == GPS_ONLY)) {
+        // GPS is available and has been zeroed, are armed, and user is not requesting baro only
+        if (baroAltitudeZeroed && (positionConfig()->altitude_source == DEFAULT)) {
```

Comment:
(in future, it would be better to replace heuristic in this branch with something physically sound)

Reply:
Yes, very true.  This is a temporary, simple approach (which works surprisingly well), to be replaced with something better.